orttraining/orttraining/eager/ort_aten.cpp

```diff
@@ -499,52 +505,44 @@ void resize_impl_ort_(
   return;
 }
 
-//#pragma region Hand-Implemented ATen Ops
+// #pragma region Hand-Implemented ATen Ops
 
 namespace aten {
 
-at::Tensor empty_memory_format(
+at::Tensor empty_strided(
   at::IntArrayRef size,
-  // *,
+  at::IntArrayRef stride,
   c10::optional<at::ScalarType> dtype_opt,
-  c10::optional<at::Layout> layout_opt,
-  c10::optional<at::Device> device_opt,
-  c10::optional<bool> pin_memory,
-  c10::optional<at::MemoryFormat> memory_format) {
-  ORT_LOG_FN(size, dtype_opt, layout_opt, device_opt, pin_memory, memory_format);
-
-  assert(dtype_opt.has_value());
-  assert(device_opt.has_value());
+  c10::optional<at::Layout> layout_opt,  // Ignored because there's no ONNX support.
+  c10::optional<at::Device> device_opt,  // Will be ORT by the time this is dispatched.
+  c10::optional<bool> pin_memory_opt) {  // Ignored because there's no ONNX support.
+  ORT_LOG_FN(size, stride, dtype_opt, layout_opt, device_opt, pin_memory_opt);
 
-  // TODO: validate options and memory format
-  // TODO: figure out how to get the correct element type.
   OrtValue ot;
+  assert(device_opt.has_value());
+  at::ScalarType dtype = c10::dtype_or_default(dtype_opt);
   auto& invoker = GetORTInvoker(*device_opt);
-  onnxruntime::Tensor::InitOrtValue(ort_scalar_type_from_aten(*dtype_opt), onnxruntime::TensorShape(size.vec()),
-                                    invoker.GetCurrentExecutionProvider().GetAllocator(0, OrtMemTypeDefault), ot);
+  onnxruntime::Tensor::InitOrtValue(ort_scalar_type_from_aten(dtype), onnxruntime::TensorShape(size.vec()),
+                                    invoker.GetCurrentExecutionProvider().GetAllocator(0, OrtMemTypeDefault), ot,
+                                    stride.vec());
   return aten_tensor_from_ort(
     std::move(ot),
     at::TensorOptions()
       .device(*device_opt)
-      .dtype(*dtype_opt));
+      .dtype(dtype));
 }
 
-at::Tensor empty_strided(at::IntArrayRef size, at::IntArrayRef stride, c10::optional<at::ScalarType> dtype_opt,
-                         c10::optional<at::Layout> layout_opt, c10::optional<at::Device> device_opt,
-                         c10::optional<bool> pin_memory_opt) {
-  ORT_LOG_FN(size, stride, dtype_opt, layout_opt, device_opt, pin_memory_opt);
+at::Tensor empty_memory_format(
```

Comment:
Do we have any test coverage for this?

Reply:
Yes, about every 3rd test or so seems to call it (when I put a print statement in it).

Reply:
Here's a stack:
(gdb) bt
#0  torch_ort::eager::aten::empty_memory_format (size=..., dtype_opt=..., layout_opt=..., device_opt=...,
    pin_memory=..., memory_format=...) at /home/lw/code/onnxruntime/orttraining/orttraining/eager/ort_aten.cpp:542
#1  0x00007fffe00e945b in c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>), &torch_ort::eager::aten::empty_memory_format>, at::Tensor, c10::guts::typelist::typelist<c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat> > >::operator()(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) (
    args#5=..., args#4=..., args#3=..., args#2=..., args#1=..., args#0=..., this=0x22d4120)
    at /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/include/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:13
#2  c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>), &torch_ort::eager::aten::empty_memory_format>, at::Tensor, c10::guts::typelist::typelist<c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat> > >, at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) (
    functor=0x22d4120, args#0=..., args#1=..., args#2=..., args#3=..., args#4=..., args#5=...)
    at /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/include/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:424
#3  0x00007fffcac1ffdc in at::_ops::empty_memory_format::redispatch(c10::DispatchKeySet, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) () from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#4  0x00007fffcae452af in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>), &at::(anonymous namespace)::empty_memory_format>, at::Tensor, c10::guts::typelist::typelist<c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat> > >, at::Tensor (c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) () from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#5  0x00007fffcac5ae30 in at::_ops::empty_memory_format::call(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>) ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#6  0x00007fffca41e0a1 in at::empty(c10::ArrayRef<long>, c10::TensorOptions, c10::optional<c10::MemoryFormat>) ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#7  0x00007fffca7b9827 in at::native::abs(at::Tensor const&) ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#8  0x00007fffcaf5cddc in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&), &at::(anonymous namespace)::(anonymous namespace)::wrapper__abs>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&> >, at::Tensor (at::Tensor const&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&) ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#9  0x00007fffcab1a4f2 in at::Tensor c10::Dispatcher::redispatch<at::Tensor, at::Tensor const&>(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&)> const&, c10::DispatchKeySet, at::Tensor const&) const [clone .isra.155] ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
--Type <RET> for more, q to quit, c to continue without paging--
#10 0x00007fffcab1a5ae in at::_ops::abs::redispatch(c10::DispatchKeySet, at::Tensor const&) ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#11 0x00007fffcbb11c8e in torch::autograd::VariableType::(anonymous namespace)::abs(c10::DispatchKeySet, at::Tensor const&) () from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#12 0x00007fffcbb1227f in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::DispatchKeySet, at::Tensor const&), &torch::autograd::VariableType::(anonymous namespace)::abs>, at::Tensor, c10::guts::typelist::typelist<c10::DispatchKeySet, at::Tensor const&> >, at::Tensor (c10::DispatchKeySet, at::Tensor const&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&) ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#13 0x00007fffcab559d0 in at::_ops::abs::call(at::Tensor const&) ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#14 0x00007fffc8a91d7c in torch::autograd::THPVariable_abs(_object*, _object*, _object*) ()
   from /home/lw/venvs/onnxruntime/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#15 0x00000000005f6929 in PyCFunction_Call ()

Reply:
I suggest adding a test using torch.empty to be specific on this functionality.